### PR TITLE
Stop gradient of the exp-normalize shift in NeuralNetworkWaveFunction

### PR DESCRIPTION
**Pull request by peracutor**
## TL;DR

A single line is added to stop the gradient of the exp-normalize shift `xs_shift` in
`NeuralNetworkWaveFunction.__call__`.  The gradient is not needed while also being the source of a latent `NaN`.

## Why stop the gradient
The exp-normalize trick is used to sum up the different determinants in `NeuralNetworkWaveFunction`. The maximum exponent (`xs_shift`) is subtracted from all exponents before summing, and added back afterwards for numerical stability. Mathematically, `log_psi` is independent of `xs_shift` and the gradient of `xs_shift` is not needed for correctness. Not computing it saves unnecessary computation. `jax.nn.logsumexp` also stops the gradient of its max.

Apart from unnecessary compute, taking the gradient of `jnp.max` (i.e. `xs_shift` here) sometimes causes a `NaN`:

`jnp.max(xs)` is differentiated as

`sum(mask * tangent) / sum(mask)` with `mask = (xs == xs.max())`

and when XLA recomputes `xs` in a different fusion than `xs.max()`, then ` xs == xs.max()` is all `False` because they don't agree bit for bit, resulting in 0/0. (Claude did heavy lifting to find this out.)

### Possibly similar problem in additive backflow
This bug could in theory also arise in the additive backflow in [nn_wave_function.py:29](https://github.com/deepqmc/deepqmc/blob/a81da0e92eef415aca91b788b6bc906664abea55/src/deepqmc/wf/nn_wave_function.py#L29) where `jnp.min` is differentiated. Here, the gradient can not just be stopped, so another fix would be necessary.

### CODE REVIEW

**Comment by zenoone** [comment]
Welcome to DeepQMC and many thanks for the PR!

I agree that the shift should not alter the gradient, and I don’t think an explicit derivation is necessary. Thanks for spotting this detail. Since deep-learning QMC is particularly prone to NaNs and numerical instabilities, I’m always happy to hear that stability can be improved “for free.”

**Review by zenoone — APPROVED** [review]
